unix/tar_unix.ml

```diff
@@ -15,60 +15,247 @@
+  let* fd = safe Unix.(openfile filename [ O_RDONLY ]) 0 in
+  let rec go t fd ?global ?data acc =
+    let* data = match data with
+      | None ->
+        let buf = Bytes.make Tar.Header.length '\000' in
+        let* () = read_complete fd buf Tar.Header.length in
+        Ok (Bytes.unsafe_to_string buf)
+      | Some data -> Ok data
+    in
+    match Tar.decode t data with
+    | Ok (t, Some `Header hdr, g) ->
+      let global = Option.fold ~none:global ~some:(fun g -> Some g) g in
+      let* acc' = f fd ?global hdr acc in
+      let* _off = seek fd (Tar.Header.compute_zero_padding_length hdr) in
+      go t fd ?global acc'
+    | Ok (t, Some `Skip n, g) ->
+      let global = Option.fold ~none:global ~some:(fun g -> Some g) g in
+      let* _off = seek fd n in
+      go t fd ?global acc
+    | Ok (t, Some `Read n, g) ->
+      let global = Option.fold ~none:global ~some:(fun g -> Some g) g in
+      let buf = Bytes.make n '\000' in
+      let* () = read_complete fd buf n in
+      let data = Bytes.unsafe_to_string buf in
+      go t fd ?global ~data acc
+    | Ok (t, None, g) ->
+      let global = Option.fold ~none:global ~some:(fun g -> Some g) g in
+      go t fd ?global acc
+    | Error `Eof -> Ok acc
+    | Error `Fatal _ as e -> e
+  in
+  Fun.protect
+    ~finally:(fun () -> safe_close fd)
+    (fun () -> go (Tar.decode_state ()) fd init)
 
-  let really_write fd buf =
-    let offset = ref 0 in
-    while !offset < String.length buf do
-      offset := !offset + with_restart Unix.write_substring fd buf !offset (String.length buf - !offset)
-    done
-end
+let unix_err_to_msg = function
+  | `Unix (e, f, s) ->
+    `Msg (Format.sprintf "error %s in function %s %s"
+            (Unix.error_message e) f s)
 
-module HeaderReader = Tar.HeaderReader(Direct)(Driver)
-module HeaderWriter = Tar.HeaderWriter(Direct)(Driver)
+let copy ~src_fd ~dst_fd len =
+  let blen = 65536 in
+  let buffer = Bytes.make blen '\000' in
+  let rec read_write ~src_fd ~dst_fd len =
+    if len = 0 then
+      Ok ()
+    else
+      let l = min blen len in
+      let* () =
+        Result.map_error
+          (function
+            | `Unix _ as e -> unix_err_to_msg e
+            | `Unexpected_end_of_file ->
+              `Msg "Unexpected end of file")
+          (read_complete src_fd buffer l)
+      in
+      let* _written =
+        Result.map_error unix_err_to_msg
+          (safe (Unix.write dst_fd buffer 0) l)
+      in
+      read_write ~src_fd ~dst_fd (len - l)
+  in
+  read_write ~src_fd ~dst_fd len
 
-include Driver
+let extract ?(filter = fun _ -> true) ~src dst =
+  let f fd ?global:_ hdr () =
+    if filter hdr then
+      match hdr.Tar.Header.link_indicator with
+      | Tar.Header.Link.Normal ->
+        let* dst =
+          Result.map_error unix_err_to_msg
+            (safe Unix.(openfile (Filename.concat dst hdr.Tar.Header.file_name)
+                          [ O_WRONLY ; O_CREAT ]) hdr.Tar.Header.file_mode)
+        in
```

Comment:
This is not super safe as the archive can contain file names such as `../../../../etc/passwd` or `/etc/passwd`. Maybe it's better to leave it up to the user to write their own function using `fold` where they can decide how to handle paths, ownership and mode?

Reply:
I'd be happy to say "extract into /my/tmp/dir", and nothing leaves that directory... thus if you have "/foo/bar" in your archive, the leading / is removed. ../ paths are ignored? I don't know for sure..

Reply:
I think that is preferable. Maybe a `unsafe_extract` could exist, but I think `Tar_unix.extract` should not be a footgun.

Reply:
`unsafe_extract` is something people can write their own with `fold` ;)